### PR TITLE
Build and sign transactions for burn/mint/set-bitcoin-wallet-address

### PR DESCRIPTION
**Pull request by jferrant**
Blocked by https://github.com/stacks-network/stacks-blockchain/issues/3647

Closes https://github.com/Trust-Machines/core-eng/issues/262
Closes https://github.com/Trust-Machines/core-eng/issues/199
Closes https://github.com/Trust-Machines/core-eng/issues/172

### CODE REVIEW

File: stacks-coordinator/src/stacks_transaction.rs [hunk 1]
```diff
@@ -18,3 +26,18 @@ pub struct StacksTransaction {
     pub postConditionMode: PostConditionMode,
     pub postConditions: LengthPrefixedList,
 }
+
+impl StacksTransaction {
+    pub fn to_blockstack_transaction(self) -> Result<BlockstackTransaction, Error> {
```
Comment (stjepangolemac) [inline stacks-coordinator/src/stacks_transaction.rs:31]: Maybe you're already aware but here are some breadcrumbs about serializing this tx before sending it to the network

https://github.com/stacks-network/stacks-blockchain/blob/73af120f2bc24ce2e4d0dc820956985dfe44eb88/src/chainstate/stacks/transaction.rs#L578-L593

**Review by stjepangolemac — APPROVED** [review]

File: stacks-coordinator/src/stacks_transaction.rs (hunk 1 above)
Comment (jferrant) [reply stacks-coordinator/src/stacks_transaction.rs:66]: Yeah I had tried doing that earlier, but as far as I can tell, it expects a hash representation of the tx, not the json blob. So not entirely sure if I have done it correctly, but fingers crossed. It seems to work. Just need to add some valid testing data.

File: stacks-coordinator/src/make_contract_call.rs [hunk 2]
```diff
@@ -87,22 +90,23 @@ impl SignedContractCallOptions {
             nonce: None,
```
Comment (netrome) [inline stacks-coordinator/src/make_contract_call.rs:90]: Not sure if this is the problem you're experiencing, but we must query the next nonce from the stacks-node and set it here.

You can get the nonce using `/v2/accounts/[Principal]`

Example response `docs/rpc/api/core-node/get-account-data.example.json`

```
{
  "balance": "0x0000000000000000000000000007a120",
  "locked": "0x0000000000000000000000000007a120",
  "unlock_height": 126,
  "nonce": 2867,
  "balance_proof": "0xabce",
  "nonce_proof": "0xabcd"
}
```

File: frost-signer/src/util.rs [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+pub fn parse_public_key(public_key: &str) -> ecdsa::PublicKey {
+    ecdsa::PublicKey::try_from(public_key)
+        .expect("failed to parse coordinator_public_key from config")
+}
+
+pub fn parse_public_keys(public_keys: &[String]) -> Vec<ecdsa::PublicKey> {
+    public_keys
+        .iter()
+        .map(|s| ecdsa::PublicKey::try_from(s.as_str()).expect("failed to parse ecdsa::PublicKey"))
+        .collect::<Vec<ecdsa::PublicKey>>()
```
Comment (sergey-shandar) [inline frost-signer/src/util.rs:12]:
```suggestion
        .collect()
```

Comment (xoloki) [reply frost-signer/src/util.rs:12]: That's my fault, I moved this code from distributed copypasta into util funds, and forgot to remove the now unnecessary generic arg to `collect`

**Comment by sergey-shandar** [comment]
@jferrant don't hesitate to fix hard coded values for `make_contract_call` `left == right` tests if needed.

File: frost-signer/src/util.rs (hunk 3 above)
Comment (sergey-shandar) [reply frost-signer/src/util.rs]: Yeah, I do this all the time. BTW, I use `.collect::<Vec<_>>()` to help the Rust type interference algorithm without specifying too much.

File: stacks-coordinator/src/make_contract_call.rs [hunk 4]
```diff
@@ -66,13 +68,16 @@ pub struct SignedContractCallOptions {
...
 impl SignedContractCallOptions {
+    #[allow(clippy::too_many_arguments)]
     pub fn new(
         contract_address: impl Into<String>,
         contract_name: impl Into<String>,
         function_name: impl Into<String>,
         function_args: &[Value],
         anchor_mode: AnchorMode,
         sender_key: impl Into<String>,
+        nonce: u64,
```
Comment (sergey-shandar) [inline stacks-coordinator/src/make_contract_call.rs:79]: Could we add a new function `with_nonce` instead? I would prefer minimize a number of parameters for functions, so don't add optional parameters to the functions signature.

**Comment by jferrant** [comment]
I pulled make_contract_call out entirely, but I could add it back and leave the original generic StacksTransaction as a potential route to explore? I am more in favour of the purely rust approach. I find it much easier to grok.

**Review by sergey-shandar — APPROVED** [review]
Awesome! No dependency on unstable and hard to debug JS code.

**Comment by jferrant** [comment]
> Awesome! No dependency on unstable and hard to debug JS code.
I created https://github.com/Trust-Machines/core-eng/issues/244 to try to decouple some of this code as you suggested :)

File: stacks-coordinator/src/stacks_wallet.rs [hunk 5]
```diff
@@ -1,84 +1,357 @@
...
+        op: &PegOutRequestOp,
+        nonce: u64,
     ) -> Result<StacksTransaction, PegWalletError> {
-        Ok(self.call("burn!".to_string())?)
+        let function_name = "burn!";
+
+        // Build the function arguments
+        let amount = Value::UInt(op.amount.into());
+        // Retrieve the address from the Message Signature
+        let pub_key = StacksPublicKey::recover_to_pubkey(op.txid.as_bytes(), &op.signature)
```
Comment (netrome) [inline stacks-coordinator/src/stacks_wallet.rs:214]: This is incorrect. The signature in the `PegOutRequestOp` signs the sha512 hash of the amount and the locking script of the peg out request transaction. I realize now that this information is not available to the coordinator... :upside_down_face: so we need to figure out how to get that information here.

Option 1: The coordinator fetches this transaction from a bitcoin node in order to retrieve the missing locking script. Ugh...
Option 2: We extend `PegOutRequestOp` in stacks-node to hold this information directly.

Thinking out loud: :thinking: 
From a security perspective both options rely on trusting information received from a running blockchain node (either Stacks or Bitcoin). Given that we'll interface with trusted nodes we host ourselves, this is fine.

I propose proceeding with Option 2. I can write up an issue for it if you think it makes sense.

**Review by netrome — CHANGES_REQUESTED** [review]
Houston we have a problem (regarding the `PegOutRequestOp`).

File: stacks-coordinator/src/stacks_wallet.rs (hunk 5 above)
Comment (jferrant) [reply stacks-coordinator/src/stacks_wallet.rs:214]: I am fine with Option 2. I feel like all information required to do something with a burn_op should be in the burn_op itself.

File: stacks-coordinator/src/config.rs [hunk 6]
```diff
@@ -22,6 +22,8 @@ pub struct Config {
     pub signer_config_path: String,
     pub start_block_height: Option<u64>,
     pub rusqlite_path: Option<String>,
+    /// The network version we are using ('mainnet' or 'testnet'). Default: 'mainnet'
+    pub network: Option<String>,
```
[the remark this answers is not in the document]
Comment (jferrant) [reply stacks-coordinator/src/config.rs:26]: Good catch! My bad for not doing a cleanup runthrough. Was in such haste to get the darn thing done :P

File: stacks-coordinator/src/stacks_wallet.rs [hunk 7]
```diff
@@ -1,84 +1,357 @@
...
-    contract_address: String,
+    contract_address: StacksAddress,
     contract_name: String,
-    sender_key: String,
+    sender_key: StacksPrivateKey,
+    version: TransactionVersion,
+    address: StacksAddress,
 }
 
 impl StacksWallet {
```
[the remark this answers is not in the document]
Comment (jferrant) [reply stacks-coordinator/src/stacks_wallet.rs:69]: I agree. Done :D

File: stacks-coordinator/src/stacks_wallet.rs (hunk 5 above)
Comment (netrome) [reply stacks-coordinator/src/stacks_wallet.rs:211]: Issue formulated: https://github.com/stacks-network/stacks-blockchain/issues/3647 Let's leave it to the Stacks-side to derive the address of the sender (pub_key) https://github.com/stacks-network/stacks-blockchain/issues/3647

File: stacks-coordinator/src/coordinator.rs [hunk 8]
```diff
@@ -127,15 +131,26 @@ pub trait Coordinator: Sized {
 // Private helper functions
 trait CoordinatorHelpers: Coordinator {
     fn peg_in(&mut self, op: stacks_node::PegInOp) -> Result<()> {
-        let _tx = self.fee_wallet().stacks_mut().build_mint_transaction(&op)?;
-        //self.stacks_node().broadcast_transaction(&tx);
+        let nonce = self
+            .stacks_node()
+            .next_nonce(self.fee_wallet().stacks().get_address())?;
```
Comment (jferrant) [inline stacks-coordinator/src/coordinator.rs:136]: @netrome is this actually correct? To always grab the nonce using the stacks wallet address?

Comment (netrome) [reply stacks-coordinator/src/coordinator.rs:136]: I think so. We ask the stacks node for the next nonce for this particular address, which it keeps track of.

Comment (soju-drinker) [reply stacks-coordinator/src/coordinator.rs:136]: Can we write this in a comment for future usage

File: stacks-coordinator/src/stacks_wallet.rs (hunk 5 above)
Comment (jferrant) [reply stacks-coordinator/src/stacks_wallet.rs]: Pulled in the change and updated the tests :) had to update the cargo file to point to "next"

File: stacks-coordinator/src/coordinator.rs [hunk 9]
```diff
@@ -217,19 +236,53 @@ impl TryFrom<Config> for StacksCoordinator {
...
+        // 1. No aggregate public key stored in persitent storage anywhere
+        // 2. no address already set in sbtc contract (get-bitcoin-wallet-address)
+        frost_coordinator.run_distributed_key_generation()?;
+        let public_key = bitcoin::PublicKey::from_slice(
+            frost_coordinator
+                .get_aggregate_public_key()?
+                .compress()
+                .as_bytes(),
+        )?;
+        let address = bitcoin::Address::p2wpkh(&public_key, bitcoin_network)?;
```
Comment (netrome) [inline stacks-coordinator/src/coordinator.rs:259]: This should be a p2tr address, since we're using Schnorr signatures.

**Review by netrome — CHANGES_REQUESTED** [review]
Good stuff. There's a p2wpkh that I believe should be a p2tr. Please update it and I'll approve.

File: stacks-coordinator/src/coordinator.rs (hunk 9 above)
Comment (jferrant) [reply stacks-coordinator/src/coordinator.rs:259]: There are quite a few places where I make this mistake. Will update.

File: stacks-coordinator/src/peg_wallet.rs [hunk 10]
```diff
@@ -35,19 +43,20 @@ pub trait BitcoinWallet {
         &self,
         op: &stacks_node::PegOutRequestOp,
     ) -> Result<bitcoin_node::BitcoinTransaction, Error>;
+    /// Gets the BTC address for the wallet
+    fn get_address(&self) -> &BitcoinAddress;
```
[the remark this answers is not in the document]
Comment (netrome) [reply stacks-coordinator/src/peg_wallet.rs:47]: @jferrant any opinions?

File: stacks-coordinator/src/coordinator.rs (hunk 9 above)
Comment (netrome) [reply stacks-coordinator/src/coordinator.rs]: I see this is resolved now :heavy_check_mark:

File: stacks-coordinator/src/peg_wallet.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (jferrant) [reply stacks-coordinator/src/peg_wallet.rs:47]: I have no opinion for or against :) will change.

**Comment by jferrant** [comment]
> I think you missed my nit comment about `get_address`. It's not a blocker for merging though.

I saw, but I did forget :P Addressed now (no pun intended...)

**Pull request merged**